### PR TITLE
[feat] DataBaseClient

### PR DESCRIPTION
**Pull request by joshuajeschek**
Co-authored-by: sar-sturm <sar-sturm@users.noreply.github.com>
Co-authored-by: TheDarkDesync <TheDarkDesync@users.noreply.github.com>
Co-authored-by: Rotwolf <Rotwolf@users.noreply.github.com>

### CODE REVIEW

File: src/test/java/github/weichware10/util/config/ZoomMapsConfigurationTest.java [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 import static org.junit.Assert.assertEquals;
 
+import java.util.Arrays;
+
 import org.junit.Test;
```
Comment (weichbot10) [inline src/test/java/github/weichware10/util/config/ZoomMapsConfigurationTest.java:7]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'org.junit.Test'

File: src/test/java/github/weichware10/util/config/CodeChartsConfigurationTest.java [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 import static org.junit.Assert.assertEquals;
 
+import java.util.Arrays;
+
 import org.junit.Test;
```
Comment (weichbot10) [inline src/test/java/github/weichware10/util/config/CodeChartsConfigurationTest.java:7]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'org.junit.Test'

File: src/main/java/github/weichware10/util/db/Configurations.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                // komplette Konfiguration zurückgeben
+                return new Configuration(configId, zoomMapsConfiguration);
+            }
+        } catch (SQLException e) {
+            Logger.error("SQLException when processing query result", e);
+            return null;
+        }
+    }
+
+    public String setConfiguration(Configuration configuration) {
```
Comment (weichbot10) [inline src/main/java/github/weichware10/util/db/Configurations.java:98]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.javadoc.MissingJavadocMethodCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing a Javadoc comment.

File: src/main/java/github/weichware10/util/db/Configurations.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        String id = null;
+
+        while (!idUnique) {
+            idUnique = true;
+            id = Util.generateId("con_", 5);
+            try {
+                Connection conn = DriverManager.getConnection(url, props);
+                Statement st = conn.createStatement();
+                if (configuration.getToolType() == ToolType.CODECHARTS) {
+                    st.executeQuery(String.format(zmQuery, id, "CODECHARTS", 1, "tst", "tst", 3.0f));
```
Comment (weichbot10) [inline src/main/java/github/weichware10/util/db/Configurations.java:128]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.sizes.LineLengthCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Line is longer than 100 characters (found 101).

File: src/main/java/github/weichware10/util/config/ZoomMapsConfiguration.java [hunk 5]
```diff
@@ -8,6 +10,16 @@
 
     // GETTER
 
+    public ZoomMapsConfiguration() {
+        // leer
+    }
+
+    public ZoomMapsConfiguration(float speed, boolean tutorial, List<String> imageUrls) {
```
Comment (weichbot10) [inline src/main/java/github/weichware10/util/config/ZoomMapsConfiguration.java:17]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.javadoc.MissingJavadocMethodCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing a Javadoc comment.

File: src/test/java/github/weichware10/util/db/TrialsTest.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+        // TODO: Wenn Überprüfung auf amount und korrekte configId in addTrials implementiert ist
+        // assertThrows(expectedThrowable, dbClient.trials.addTrials("nonExisitingConfigId", 3));
+        // assertThrows(expectedThrowable, dbClient.trials.addTrials(configIdZm, -1));
+        // assertThrows(expectedThrowable, dbClient.trials.addTrials("nonExisitingConfigId", -1));
+    }
+
+    @Test
+    public void setTrialShouldWork() {
+
```
Comment (weichbot10) [inline src/test/java/github/weichware10/util/db/TrialsTest.java:53]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.regexp.RegexpSinglelineCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Line has trailing whitespaces.

Comment (weichbot10) [inline src/test/java/github/weichware10/util/db/TrialsTest.java:53]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.regexp.RegexpSinglelineCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Line has trailing whitespaces.

Comment (weichbot10) [inline src/test/java/github/weichware10/util/db/TrialsTest.java:53]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.regexp.RegexpSinglelineCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Line has trailing whitespaces.

File: src/main/java/github/weichware10/util/data/TrialData.java [hunk 7]
```diff
@@ -3,13 +3,15 @@
...
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import org.joda.time.DateTime;
 
 /**
  * Stores the TrialData for the different tools internally.
  *
- * <p>also used to transfer Data
+ * <p>
```
Comment (weichbot10) [inline src/main/java/github/weichware10/util/data/TrialData.java:13]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocParagraphCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br><p> tag should be placed immediately before the first word, with no space after.

File: src/main/java/github/weichware10/util/data/DataPoint.java [hunk 8]
```diff
@@ -2,6 +2,8 @@
 
 import java.util.Arrays;
 
+import github.weichware10.util.Logger;
```
Comment (weichbot10) [inline src/main/java/github/weichware10/util/data/DataPoint.java:5]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'github.weichware10.util.Logger'

Comment (weichbot10) [inline src/main/java/github/weichware10/util/data/DataPoint.java:5]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Wrong lexicographical order for 'github.weichware10.util.Logger' import. Should be before 'java.util.Arrays'.

File: src/test/java/github/weichware10/util/db/TrialsTest.java [hunk 9]
```diff
@@ -0,0 +1,309 @@
+package github.weichware10.util.db;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+import github.weichware10.util.Logger;
+import github.weichware10.util.Enums.ToolType;
```
Comment (weichbot10) [inline src/test/java/github/weichware10/util/db/TrialsTest.java:9]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Wrong lexicographical order for 'github.weichware10.util.Enums.ToolType' import. Should be before 'github.weichware10.util.Logger'.

File: src/test/java/github/weichware10/util/db/TrialsTest.java [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+import github.weichware10.util.config.CodeChartsConfiguration;
+import github.weichware10.util.config.Configuration;
+import github.weichware10.util.config.ZoomMapsConfiguration;
+import github.weichware10.util.data.DataPoint;
+import github.weichware10.util.data.TrialData;
+import io.github.cdimascio.dotenv.Dotenv;
+
+import java.util.ArrayList;
```
Comment (weichbot10) [inline src/test/java/github/weichware10/util/db/TrialsTest.java:17]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'java.util.ArrayList'

File: src/test/java/github/weichware10/util/db/TrialsTest.java [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+import org.joda.time.DateTime;
```
Comment (weichbot10) [inline src/test/java/github/weichware10/util/db/TrialsTest.java:21]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'org.joda.time.DateTime'

File: src/test/java/github/weichware10/util/db/TrialsTest.java [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+                new DateTime(2021, 12, 13, 14, 0, 0), "Ja", new ArrayList<DataPoint>());
+        trialData3.addDataPoint(new int[] { 3, 3 }, 3.0f);
+        dbClient.trials.setTrial(trialData3);
+
+        String configIdZm2 = dbClient.configurations.setConfiguration(zoomConfig);
+        List<String> trialsZm2 = dbClient.trials.addTrials(configIdZm2, 1);
+
+        TrialData trialData7 = new TrialData(ToolType.ZOOMMAPS, trialsZm2.get(0), configIdZm2,
+                new DateTime(2021, 12, 10, 14, 0, 0), "Ja", new ArrayList<DataPoint>());
+                trialData7.addDataPoint(new int[] { 4, 4 }, 4.0f);
```
Comment (weichbot10) [inline src/test/java/github/weichware10/util/db/TrialsTest.java:207]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.indentation.IndentationCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>'method def' child has incorrect indentation level 16, expected level should be 8.

File: src/test/java/github/weichware10/util/db/TrialsTest.java [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+                new DateTime(2021, 12, 11, 22, 0, 0), "Ja", new ArrayList<DataPoint>());
+        trialData6.addDataPoint(new int[] { 2, 2 }, (new int[] { 2, 2 }));
+        dbClient.trials.setTrial(trialData6);
+
+        String configIdCc2 = dbClient.configurations.setConfiguration(codeConfig);
+        List<String> trialsCc2 = dbClient.trials.addTrials(configIdCc2, 1);
+
+        TrialData trialData8 = new TrialData(ToolType.CODECHARTS, trialsCc2.get(0), configIdCc2,
+                new DateTime(2021, 12, 9, 14, 0, 0), "Ja", new ArrayList<DataPoint>());
+                trialData8.addDataPoint(new int[] { 4, 4 }, new int[] { 4, 4 });
```
Comment (weichbot10) [inline src/test/java/github/weichware10/util/db/TrialsTest.java:264]: ⚠️ **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.indentation.IndentationCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>'method def' child has incorrect indentation level 16, expected level should be 8.

File: src/main/java/github/weichware10/util/config/Configuration.java [hunk 14]
```diff
@@ -9,44 +9,62 @@
...
         ; // ohne Initialisierungen
     }
 
     /**
-     * allgemeiner Speicher-Ort für Einstellungen.
-     * (Constructor mit Übergabe von non-default-Tools)
+     * Konstruktor für Configuration vom Typ CodeChartsConfiguration.
+     *
+     * @param configId - configId der übergebenen Konfiguration
+     * @param codeChartsConfiguration - Konfiguration von CodeCharts
```
Comment (sar-sturm) [inline src/main/java/github/weichware10/util/config/Configuration.java:30]: Fehlt hier nicht das Param von question? Genauso beim nächsten

File: src/main/java/github/weichware10/util/Tutorial.java [hunk 15]
```diff
@@ -46,8 +46,8 @@ private boolean getConfigState() {
         switch (toolType) {
             case CODECHARTS:
                 return configClient.getConfig().getCodeChartsConfiguration().getTutorial();
-            case EYETRACKING:
-                return configClient.getConfig().getEyeTrackingConfiguration().getTutorial();
+            // case EYETRACKING:
+            //     return configClient.getConfig().getEyeTrackingConfiguration().getTutorial();
             default: // ZOOMMAPS
```
Comment (Rotwolf) [inline src/main/java/github/weichware10/util/Tutorial.java:51]: Maybe Deffensiver Programmeiren?

Comment (joshuajeschek) [reply src/main/java/github/weichware10/util/Tutorial.java:51]: ist jetzt erstmal egal :)

**Comment by joshuajeschek** [comment]
DataBaseClient+Util: @Rotwolf
Configurations+DataPoints: @Nilrei
Trials.java: @sar-sturm @grenoui11e
Tests nicht vergessen

**Review by Rotwolf — APPROVED** [review]
Für meinen Teil passt alles ^^

File: src/main/java/github/weichware10/util/db/Trials.java [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+     */
+    public TrialData getTrial(String trialId) {
+        final String queryF = """
+                SELECT t.configid, c.tooltype, t.starttime, t.answer
+                FROM %s.trials AS t, %s.configurations AS c
+                WHERE t.configid LIKE c.configid AND t.trialid LIKE '%s'""";
+        final String query = String.format(
+                    queryF,
+                    dataBaseClient.schema,
+                    dataBaseClient.schema,
```
Comment (sar-sturm) [inline src/main/java/github/weichware10/util/db/Trials.java:73]: soll dataBaseClient.schema 2 mal drinstehen?

File: src/main/java/github/weichware10/util/db/Trials.java [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+     * @param amount   - Anzahl der Versuche
+     * @return Liste mit vergebenen trialIDs
+     */
+    public List<String> add(String configId, int amount) {
+        // Überprüft, ob amount <= 0 ist
+        // oder Verfügbarkeit der Konfiguration mit der configId false ist
+        if (amount <= 0 || !dataBaseClient.configurations.getAvailability(configId)) {
+            return null;
+        }
+        final String query = """
```
Comment (grenoui11e) [inline src/main/java/github/weichware10/util/db/Trials.java:219]: Fehlt hier nicht noch die final String queryF = """[...]""" in der der die SQL Anfrage steht?
Sodass die final String query = String.format(...); macht?

**Review by grenoui11e — APPROVED** [review]

**Review by sar-sturm — APPROVED** [review]
Bis auf die offenen Fragen oben sieht alles gut aus

File: src/main/java/github/weichware10/util/db/Trials.java (hunk 16 above)
Comment (joshuajeschek) [reply src/main/java/github/weichware10/util/db/Trials.java:73]: ja, schau mal weiter oben die Variable queryF, dort muss das 2 Mal eingefügt werden. Man könnte das zweite `%s` aber auch durch `%>s` ersetzen, um das vom vorherigen zu übernehmen

File: src/main/java/github/weichware10/util/db/Trials.java (hunk 17 above)
Comment (joshuajeschek) [reply src/main/java/github/weichware10/util/db/Trials.java:219]: das wird erst unten gemacht, da ja das ganze mit verschiedenen trialIds, mehrmals gemacht wird. Ich kann den Variablenname aber zu queryF ändern, damit das offensichtlicher ist

**Pull request merged**